2020/day-4-passport-processing/index.ts

```diff
@@ -0,0 +1,75 @@
+// byr (Birth Year) - four digits; at least 1920 and at most 2002.
+// iyr (Issue Year) - four digits; at least 2010 and at most 2020.
+// eyr (Expiration Year) - four digits; at least 2020 and at most 2030.
+// hgt (Height) - a number followed by either cm or in:
+//     If cm, the number must be at least 150 and at most 193.
+//     If in, the number must be at least 59 and at most 76.
+// hcl (Hair Color) - a # followed by exactly six characters 0-9 or a-f.
+// ecl (Eye Color) - exactly one of: amb blu brn gry grn hzl oth.
+// pid (Passport ID) - a nine-digit number, including leading zeroes.
+// cid (Country ID) - ignored, missing or not.
+
+const parseList = (list: string) => {
+  return list
+    .trim()
+    .split(/\n\n/)
+    .map(entry => entry.split(/\s+/).map(x => x.split(':')))
+    .map(Object.fromEntries)
+}
+
+type Passport = {
+  byr?: string
+  iyr?: string
+  eyr?: string
+  hgt?: string
+  hcl?: string
+  ecl?: string
+  pid?: string
+  cid?: string
+}
+
+type Maybe<T> = T | null | undefined
+
+const isWithinRange = (value: string | number, start: string | number, end: string | number) =>
+  Number(value) >= start && Number(value) <= end
+
+const isValidBirthYear = (byr: Maybe<string | number>) => !!byr && isWithinRange(byr, 1920, 2002)
+
+const isValidIssueYear = (iyr: Maybe<string | number>) => !!iyr && isWithinRange(iyr, 2010, 2020)
+
+const isValidExpirationYear = (eyr: Maybe<string | number>) => !!eyr && isWithinRange(eyr, 2020, 2030)
+
+const isValidHeight = (hgt: Maybe<string>) => {
+  const [_, value, unit] = hgt?.match(/([0-9]+)(cm|in)$/) || []
+
+  if (unit === 'cm') {
+    return isWithinRange(value, 150, 193)
+  }
+
+  if (unit === 'in') {
+    return isWithinRange(value, 59, 76)
+  }
+
+  return false
+}
+
+const isValidHairColor = (hcl: Maybe<string>) => !!hcl?.match(/^#[0-9a-f]{6}/gi)
```

Comment:
The use of `gi` in your regex here feels a bit off to me.

- `g` stands for global matches, i.e match all occurrences of the match. I don't think we need this with this test as we simply care if the string matches yes or no. Not whether it matches multiple times, we aren't using the groups returned here.

- `i` stands for case insensitive. The fact that you have used `[a-f]` implies this is case sensitive. The pattern to use here if you don't care about case sensitivity is `[a-zA-Z0-9]`.

```suggestion
const isValidHairColor = (hcl: Maybe<string>) => !!hcl?.match(/^#[0-9a-f]{6}/)
```

Also is there a reason that you used `a-f` instead of `a-z`?

Reply:
Hmm, I agree on the `g` flag - that can go. I feel like the point of the `i` flag though is so that you don't have to write the same thing in different cases (i.e. `a-zA-Z`).

`a-f` is because hex only use these characters - `#bada55` is valid, but `#badass` isn't